aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -14,14 +14,93 @@ static constexpr int64_t kILP = 4;
 static constexpr int64_t kChunkSize = 65536;
 static constexpr int64_t kBlockSize = 512;
 
-// TODO(crcrpar): Add `n>5` for `low prec params & their higher prec copy`
-// TensorListMetadata has to be < 4KB - the limit for kernel launch argument
-static constexpr int depth_to_max_tensors[5] = {110, 64, 48, 36, 30};
-static constexpr int depth_to_max_blocks[5] = {320, 320, 320, 320, 320};
-static constexpr int depth_to_max_tensors_scalarlist[5] = {96, 64, 48, 36, 30};
-static constexpr int depth_to_max_tensors_scalarlist_of_complex_double[2] = {
-    72,
-    60};
+// NOTE: [32KB kernel argument size support]
+// 32KB kernel argument size support is available only when CUDART_VERSION >=
+// 12010 and the driver version is >= 530. Only the former condition can be
+// checked at compile time. This implies:
+//
+// - If CUDART_VERSION < 12010, kernels for the 32KB kernel argument size will
+// not be built.
+//
+// - If CUDART_VERSION >= 12010, kernels for both 4KB and 32KB kernel argument
+// sizes will be built. However, due to CUDA's minor version compatibility,
+// even when CUDART_VERSION >= 12010, the driver may not support 32KB.
+// Therefore, a runtime check is necessary to determine the appropriate kernel
+// to dispatch.
+//
+// - TODO(yifu): once there's a CUDART version that is not compatible with any
+// driver version below 530, we can determine at compile time to not compile
+// the kernels for 4KB kernel argument size.
+//
+// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
+bool supports_large_kernel_arg() {
+#if !defined(USE_ROCM) && defined(CUDART_VERSION) && CUDART_VERSION >= 12010
+  static std::optional<bool> supports_large_kernel_arg_ = std::nullopt;
+  if (!supports_large_kernel_arg_.has_value()) {
```

Comment:
looks like a noob q: when would line 39 ever be False if line 38 sets it to null? Is `supports_large_kernel_arg_1 accessible from outside this function?

Reply:
OOOOH i missed the static. Very cool. It has the same effect as function caching then.